### DISCUSSION TITLE
Can I ask for viewport visibility for a widget attached to a form field with `@@=`?

### DISCUSSION THREAD

**Discussion by amano-kenji** [jtdaugherty/brick#489] [Q&A] [answered]
It seems there is no way to easily do so for now.

**Comment by jtdaugherty** [comment]
The form already does that for you: it makes a visibility request for the focused field. If you put the whole form in a viewport, then the focused field should always come into view. Have you tried that?

**Reply by amano-kenji** [reply]
Is `formFieldRenderHelper` included in visibility request by brick?

**Reply by jtdaugherty** [reply]
No, the visibility request is made for the field itself.

**Reply by jtdaugherty** [reply]
(If that isn't what you need, you can always make the render helper issue a visibility request.)

**Reply by amano-kenji** [reply]
Can I augment a specific field rather than `FormFieldState`?

**Reply by jtdaugherty** [reply]
I'm not sure what you mean. Can you say more about what you're trying to accomplish? Perhaps give a code sample and UI screenshot to describe what you want to do?

**Reply by amano-kenji** [reply]
If I add a label to a text field as below
```
(txt "label" <=>) @@= editTextField ...
```
`label` is not included in the visibility request made by brick for editTextField.

**Reply by jtdaugherty** [reply]
In that case you'll need to make your own visibility request, e.g. `((visible Foo $ txt "label") <=>) @@= editTextField ...`.

**Reply by amano-kenji** [reply]
I want to make visibility request  for `txt "label"` only if `editTextField` is in focus.

There can be many editTextFields from top to bottom, and they may not fit vertically.

I'm looking to augment `formFieldRender` of `FormField` so that the added label is included in the visibility request automatically by brick.

**Reply by jtdaugherty** [reply, marked as answer]
> I want to make visibility request for txt "label" only if editTextField is in focus.

That isn't possible with the current form API.

**Reply by amano-kenji** [reply]
You are right. I tried this
```
labelAboveEditField :: Widget n -> (s -> FormFieldState s e n)
  -> s -> FormFieldState s e n
labelAboveEditField label fld s = let
  ffs@FormFieldState { formFields=[editField] } = fld s
  FormField { formFieldRender=ffr } = editField
  in ffs { formFields = [ editField { formFieldRender = \foc b -> label <=> ffr foc b } ] }
```
But, this was not possible because `FormFieldState` cannot be manipulated from outside. If this was possible, brick would automatically make visibility request for label and the edit field.

**Reply by amano-kenji** [reply]
To add a label along with an edit field, I would have to write a new function that assembles both during construction of `FormFieldState`.

**Reply by jtdaugherty** [reply]
There is now support for this in 2.2.

**Reply by amano-kenji** [reply]
Cool.
